### ISSUE TITLE
Too much destruction!

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5763]
Migrated from [rt.perl.org#15273](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=15273) (status was 'resolved')

Searchable as RT15273$

**Comment by p5pRT** [comment]
### From @lizmat

#### Created by @lizmat



Take the following program in 5\.8\.0&#8203;:

== toomuch ==============================================================
use threads;
use warnings;

my $cloned = 0;

my $foo = Foo\->new;
threads\->new\( sub \{ my $otherfoo = Foo\->new \} \)\->join;
warn "now done with thread\\n";

package Foo;
sub new \{
&nbsp;   warn "Creating with cloned = $cloned\\n";
&nbsp;   bless \{cloned => $cloned\}\,shift;
\}
sub CLONE \{ $cloned\+\+ \}
sub DESTROY \{ warn "DESTROYing with cloned = $cloned = $\_\[0\] = 
$\_\[0\]\->\{'cloned'\}\\n" \}
=========================================================================

&nbsp;  $ perl toomuch
&nbsp;  Creating with cloned = 0
&nbsp;  Creating with cloned = 1
&nbsp;  DESTROYing with cloned = 1 = Foo=HASH\(0x8232910\) = 1
&nbsp;  DESTROYing with cloned = 1 = Foo=HASH\(0x8207870\) = 0
&nbsp;  Use of uninitialized value in concatenation \(\.\) or string at toomuch line 17\.
&nbsp;  Use of uninitialized value in concatenation \(\.\) or string at toomuch line 17\.
&nbsp;  DESTROYing with cloned = 1 =  =
&nbsp;  now done with thread
&nbsp;  DESTROYing with cloned = 0 = Foo=HASH\(0x8125ed4\) = 0
&nbsp;  Use of uninitialized value in concatenation \(\.\) or string at toomuch line 
17 during global destruction\.
&nbsp;  DESTROYing with cloned = 1 = HASH\(0x826f364\) =


It looks to me that DESTROY is called 2 times too many\.

The first DESTROY \(1=1\) is the $otherfoo object being destroyed when the 
thread ends\.

The second DESTROY \(1=0\) is the cloned $foo object being destroyed when the 
thread ends\.

The third DESTROY \(1=\) I have no idea why this is happening\, there is 
definitely no object there anymore\.

The fourth DESTROY \(0=0\) is the original $foo object being destroyed at the 
end of the program\.

The fifth DESTROY I have no idea why this is happening\, there is still a 
hash but is different from any of the above and it is not blessed anymore\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=high

Site configuration information for perl v5.8.0:

Configured by liz at Sat Jul 20 22:09:02 CEST 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
   Platform:
     osname=linux, osvers=2.4.18, archname=i686-linux-thread-multi
     uname='linux echt.dijkmat.nl 2.4.18 #8 smp mon mar 25 22:28:36 cet 
2002 i686 unknown '
     config_args='-de -Dusethreads'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define 
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing 
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 
-I/usr/include/gdbm',
     optimize='-O2',
     cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing 
-I/usr/local/include -I/usr/include/gdbm'
     ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-98)', 
gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=4, prototype=define
   Linker and Libraries:
     ld='cc', ldflags =' -L/usr/local/lib'
     libpth=/usr/local/lib /lib /usr/lib
     libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
     perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
     libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
     gnulibc_version='2.2.4'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
     cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.0:
     /usr/local/lib/perl5/5.8.0/i686-linux-thread-multi
     /usr/local/lib/perl5/5.8.0
     /usr/local/lib/perl5/site_perl/5.8.0/i686-linux-thread-multi
     /usr/local/lib/perl5/site_perl/5.8.0
     /usr/local/lib/perl5/site_perl/5.7.3
     /usr/local/lib/perl5/site_perl
     .


Environment for perl v5.8.0:
     HOME=/home/liz
     LANG=en_US
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/liz/bin
     PERL_BADLANG (unset)
     SHELL=/bin/bash






```
</details>

**Comment by p5pRT** [comment]
### From @ysth

In article \<rt\-15273\-30860\.15\.5115747543@&#8203;perl>\,
Elizabeth Mattijsen \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
>threads\->new\( sub \{ my $otherfoo = Foo\->new \} \)\->join;

Does it change anything if you say sub \{ my $otherfoo = Foo\->new; return \}
instead?  Just curious\.\.\.

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From arthur@contiller.se

This is solved by \#18419

What happened was that the package Foo was cloned as part of the object
and we ended up with two mains and two Foos and that led to some great
confusion with the refcounts\.

Now instead we see if the stash already exists and use instead of
cloning it\. New stashes will be cloned if needed\.

This should cut back drasticly on memory usage\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'open' to 'resolved'
